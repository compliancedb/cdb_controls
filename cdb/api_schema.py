class ApiSchema:
    @staticmethod
    def url_for_owner_projects(host, project_data):
        return host + '/api/v1/projects/' + project_data["owner"] + "/"

    @staticmethod
    def url_for_project(host, project_data):
        return ApiSchema.url_for_owner_projects(host, project_data) + project_data["name"]

    @staticmethod
    def url_for_releases(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/releases/'

    @staticmethod
    def url_for_commit(host, project_data, commit):
        return ApiSchema.url_for_project(host, project_data) + '/commits/' + commit

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_artifact(host, project_data, sha256_digest):
        return ApiSchema.url_for_artifacts(host, project_data) + sha256_digest

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_release(host, project_data, release_number):
        return ApiSchema.url_for_releases(host, project_data) + release_number