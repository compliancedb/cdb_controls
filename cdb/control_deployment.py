def control_deployment_approved(approvals):
    for approval in approvals:
        if approval["state"] == "APPROVED":
            return True
    return False
