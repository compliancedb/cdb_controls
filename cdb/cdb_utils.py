import getopt
import json
import os
import sys
import docker
import requests as req
from junitparser import JUnitXml
from requests.auth import HTTPBasicAuth

from cdb.settings import CDB_SERVER

CMD_HELP = 'ensure_project.py -p <project.json>'


def project_exists_in_cdb(project_data, projects):
    exists = False
    for project in projects:
        if project["name"] == project_data["name"]:
            exists = True
    return exists


def get_project_list_from_cdb(projects_url):
    resp = req.get(projects_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    projects = resp.json()
    return projects


def get_project_from_cdb(project_url):
    resp = req.get(project_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    project = resp.json()
    return project


def load_project_configuration(json_data_file):
    project_data = json.load(json_data_file)
    return project_data


def create_artifact(api_token, host, project_config_file, sha256, filename, description, git_commit, commit_url, build_url,
                    is_compliant):
    project_data = load_project_configuration(project_config_file)

    '''
    curl -H 'Content-Type: application/json' \
     -X PUT \
     -d '{"sha256": "'"$3"'", "filename": "'"$4"'", "description": "'"$5"'", "git_commit": "'"$6"'", "commit_url": "'"$7"'", "build_url": "'"$8"'", "is_compliant": "true"}' \
    http://server:8001/api/v1/projects/$1/$2/artifacts/
    '''

    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url,
        "is_compliant": is_compliant
    }
    url = url_for_artifacts(host, project_data)
    put_payload(create_artifact_payload, url, api_token)


def put_payload(payload, url, api_token):
    headers = {"Content-Type": "application/json"}
    print("Putting this payload:")
    print(json.dumps(payload, sort_keys=True, indent=4))
    print("To url: " + url)
    resp = req.put(url, data=json.dumps(payload), headers=headers, auth=HTTPBasicAuth(api_token, 'unused'))
    print(resp.text)


def url_for_artifacts(host, project_data):
    return url_for_project(host, project_data) + project_data["name"] + '/artifacts/'


def add_evidence(api_token, host, project_file_contents, sha256_digest, evidence):
    project_data = load_project_configuration(project_file_contents)
    url_for_artifact = url_for_artifacts(host, project_data) + sha256_digest

    put_payload(evidence, url_for_artifact, api_token)


def url_for_project(host, project_data):
    return host + '/api/v1/projects/' + project_data["owner"] + '/'


def rchop(thestring, ending):
    if thestring.endswith(ending):
        return thestring[:-len(ending)]
    return thestring


def parse_cmd_line():
    project_file = ""
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hp:", ["project="])
    except getopt.GetoptError:
        print(CMD_HELP)
        sys.exit(2)
    for opt, arg in opts:
        print(opt + ": " + arg)
        if opt == '-h':
            print(CMD_HELP)
            sys.exit()
        elif opt in ("-p", "--project"):
            project_file = arg
    return project_file


def get_image_details():
    docker_image = os.getenv('CDB_DOCKER_IMAGE', "NO_DOCKER_IMAGE_FOUND")
    sha256_digest = os.getenv('CDB_ARTIFACT_SHA', None)
    if sha256_digest is None:
        client = docker.from_env()
        print("Inspecting docker image for sha256Digest")
        image = client.images.get(docker_image)
        sha256_digest = image.attrs["RepoDigests"][0].split(":")[1]
    else:
        print("Docker image digest found from environment variable")

    print("IMAGE DIGEST: " + str(image.attrs["RepoDigests"][0]))
    return docker_image, sha256_digest


def env_is_compliant():
    return os.getenv('CDB_IS_COMPLIANT', "FALSE") == "TRUE"


def is_compliant_suite(junit_xml):
    if junit_xml.failures != 0:
        return False, "Tests contain failures"
    if junit_xml.errors != 0:
        return False, "Tests contain errors"
    return True, "All tests passed"


def load_test_results(file_path):
    test_xml = JUnitXml.fromfile(file_path)
    return test_xml


def is_compliant_test_results(file_path):
    """
    This parses a junit xml file to determine if there are any errors or failures

    return:
        A tuple, with is_compliant, plus a message string
    """
    test_xml = load_test_results(file_path)
    if test_xml._tag == "testsuites":
        for suite in test_xml:
            suite_is_compliant, message = is_compliant_suite(suite)
            if not suite_is_compliant:
                return suite_is_compliant, message
        # every test suite passed, so return True
        return True, "All tests passed"
    if test_xml._tag == "testsuite":
        return is_compliant_suite(test_xml)
    return False, "Could not find test suite(s)"


def control_junit():
    print("Publish evidence to ComplianceDB")
    junit_results_path = "/data/junit/junit.xml"

    is_compliant, message = is_compliant_test_results(junit_results_path)
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "junit")
    description = "JUnit results xml verified by compliancedb/cdb_controls: " + message
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def put_evidence():
    print("Publish evidence to ComplianceDB")

    is_compliant = env_is_compliant()
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "EVIDENCE_TYPE_UNDEFINED")
    description = os.getenv('CDB_DESCRIPTION', "UNDEFINED")
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def send_evidence(evidence):
    print(evidence)
    project_file = parse_cmd_line()
    with open(project_file) as project_file_contents:
        _docker_image_unused, sha256_digest = get_image_details()
        api_token = os.getenv('CDB_API_TOKEN', 'NO_API_TOKEN_DEFINED')
        host = os.getenv('CDB_HOST', CDB_SERVER)
        add_evidence(api_token, host, project_file_contents, sha256_digest, evidence)


def build_evidence_dict(is_compliant, evidence_type, description, build_url):
    evidence = {"evidence_type": evidence_type, "contents": {
        "is_compliant": is_compliant,
        "url": "",
        "description": ""
    }}
    evidence["contents"]["description"] = description
    evidence["contents"]["url"] = build_url
    return evidence
