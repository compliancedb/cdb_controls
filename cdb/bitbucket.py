import json
import os
import sys

import requests

from cdb.cdb_utils import build_evidence_dict, send_evidence


def put_bitbucket_pull_request(project_config_file):

    pull_requests = get_pull_request_for_current_commit()

    print("Found pull request info: " + str(pull_requests))
    evidence = build_evidence_dict(
        is_compliant=True,
        evidence_type="pull_request",
        description="Bitbucket pull request",
        build_url=pull_requests[0]["pullRequestURL"])
    evidence["contents"]["source"] = pull_requests
    send_evidence(project_config_file, evidence)


def get_pull_request_for_current_commit():
    workspace = os.getenv('BITBUCKET_WORKSPACE', None)
    repository = os.getenv('BITBUCKET_REPO_SLUG', None)
    commit = os.getenv('BITBUCKET_COMMIT', None)
    user = os.getenv('BITBUCKET_USER', None)
    password = os.getenv('BITBUCKET_PWD', None)
    pull_requests = get_pull_requests_from_bitbucket_api(
        workspace=workspace, repository=repository, commit=commit, username=user, password=password)
    return pull_requests


def get_pull_requests_from_bitbucket_api(workspace, repository, commit, username, password):
    pull_requests_evidence = []

    url = f"https://api.bitbucket.org/2.0/repositories/{workspace}/{repository}/commit/{commit}/pullrequests"
    r = requests.get(url, auth=(username, password))
    if r.status_code == 200:
        pull_requests_json = json.loads(r.text)
        pull_requests = pull_requests_json["values"]
        for pr in pull_requests:
            pr_evidence = {}

            pr_html_url = pr['links']['html']['href']
            pr_api_url = pr['links']['self']['href']

            pr_evidence['pullRequestMergeCommit'] = commit
            pr_evidence['pullRequestURL'] = pr_html_url

            pr_evidence = get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password)

            pull_requests_evidence.append(pr_evidence)

    elif r.status_code == 202:
        print("Repository pull requests are still being indexed, please retry.")
        sys.exit(1)
    elif r.status_code == 404:
        print("Repository does not exists or pull requests are not indexed. Please make sure Pull Request Commit Links app is installed")
        sys.exit(2)
    else:
        print(f"Exception occured in fetching pull requests. Http return code is {r.status_code}")
        sys.exit(3)

    return pull_requests_evidence


def get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password):
    r = requests.get(pr_api_url, auth=(username, password))
    if r.status_code == 200:
        pr_json = json.loads(r.text)
        pr_evidence['pullRequestState'] = pr_json['state']
        participants = pr_json['participants']
        approvers = ""
        if len(participants) > 0:
            for participant in participants:
                if participant['approved']:
                    approvers = approvers + participant['user']['display_name'] + ","
            approvers = approvers[:-1]
            pr_evidence['approvers'] = approvers
        else:
            print("No approvers found")
    else:
        print("Error occured in fetching pull request details. Please review repository permissions.")
    return pr_evidence





