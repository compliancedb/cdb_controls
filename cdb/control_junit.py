#!/usr/bin/env python

from cdb.cdb_utils import control_junit

if __name__ == '__main__':
    control_junit()
