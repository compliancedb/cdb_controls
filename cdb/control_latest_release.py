#!/usr/bin/env python

from cdb.cdb_utils import control_latest_release

if __name__ == '__main__':
    control_latest_release()
