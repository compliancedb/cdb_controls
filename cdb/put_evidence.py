#!/usr/bin/env python

from cdb.cdb_utils import put_evidence

if __name__ == '__main__':
    put_evidence()
