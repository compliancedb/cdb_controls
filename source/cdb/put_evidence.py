#!/usr/bin/env python
import os

from cdb.cdb_utils import parse_cmd_line, env_is_compliant, load_user_data, build_evidence_dict, send_evidence


def put_evidence(project_file):
    print("Publish evidence to ComplianceDB")

    is_compliant = env_is_compliant()
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "EVIDENCE_TYPE_UNDEFINED")
    description = os.getenv('CDB_DESCRIPTION', "UNDEFINED")
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")
    user_data = load_user_data()

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url, user_data)
    send_evidence(project_file, evidence)


if __name__ == '__main__':
    project_file = parse_cmd_line()
    put_evidence(project_file)
