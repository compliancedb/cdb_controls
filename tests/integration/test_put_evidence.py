from cdb.put_evidence import put_evidence

import docker
from pytest import raises
from tests.utils import AutoEnvVars, CDB_DRY_RUN, verify_approval, auto_reading


def test_all_env_vars_uses_CDB_ARTIFACT_SHA(capsys):
    env = {
        "CDB_HOST": "http://test.compliancedb.com",
        "CDB_API_TOKEN": "5199831f4ee3b79e7c5b7e0ebe75d67aa66e79d4",
        "CDB_ARTIFACT_SHA": "a7cdaef69c676c2466571d3233380d559ccc2032b258fc5e73f99a103db462ef",
        "CDB_IS_COMPLIANT": "TRUE",
        "CDB_EVIDENCE_TYPE": "coverage",
        "CDB_DESCRIPTION": "branch coverage",
        "CDB_BUILD_NUMBER": "19",
        "CDB_CI_BUILD_URL": "https://gitlab/build/1456",
    }
    set_env_vars = {}
    with AutoEnvVars({**CDB_DRY_RUN, **env}, set_env_vars):
        put_evidence("tests/integration/test-pipefile.json")
    verify_approval(capsys, ["out"])


def test_when_no_env_vars_raises_DockerException(capsys):
    set_env_vars = {}
    with AutoEnvVars(CDB_DRY_RUN, set_env_vars), \
            raises(docker.errors.DockerException), \
            auto_reading(capsys):
        put_evidence("tests/integration/test-pipefile.json")


def test_when_neither_image_nor_sha_env_var_defined_raises_DockerException(capsys):
    env = {
        "CDB_EVIDENCE_TYPE": "test",
        "CDB_DESCRIPTION": "integration test",
        "CDB_CI_BUILD_URL": "integration test",
        "CDB_IS_COMPLIANT": "TRUE",
    }
    set_env_vars = {}
    with AutoEnvVars({**CDB_DRY_RUN, **env}, set_env_vars), \
            raises(docker.errors.DockerException), \
            auto_reading(capsys):
        put_evidence("tests/integration/test-pipefile.json")
