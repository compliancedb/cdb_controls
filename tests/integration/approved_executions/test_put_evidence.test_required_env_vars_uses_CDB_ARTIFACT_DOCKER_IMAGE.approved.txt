Publish evidence to ComplianceDB
{'evidence_type': 'unit_test', 'contents': {'is_compliant': True, 'url': 'https://gitlab/build/1956', 'description': 'branch coverage'}}
IMAGE DIGEST: bbcdaef69c676c2466571d3233380d559ccc2032b258fc5e73f99a103db462ef
Putting this payload:
{
    "contents": {
        "description": "branch coverage",
        "is_compliant": true,
        "url": "https://gitlab/build/1956"
    },
    "evidence_type": "unit_test"
}
To this url: https://app.compliancedb.com/api/v1/projects/compliancedb/cdb-controls-test-pipeline/artifacts/bbcdaef69c676c2466571d3233380d559ccc2032b258fc5e73f99a103db462ef
DRY RUN: Put not sent
